src/index.ts

```diff
@@ -13,49 +13,80 @@
+
+// Returns value of metadata field.
+// Throws error if there is a problem accessing metadata API.
+function getField(field: string): Promise<string> {
+  return new Promise<string>((resolve, reject) => {
+    gcpMetadata.instance(
+        field, (err: Error, response: any, metadata: string) => {
+          err ? reject(err) : resolve(metadata);
+        });
+  });
+}
 
 // initConfig sets unset values in the configuration to the value retrieved from
-// environment variables, metadata, or the default value specified in
+// environment variables, metadata, or the default values specified in
 // defaultConfig.
-// Returns rejected promise if value that must be set cannot be initialized.
-async function initConfig(config: Config): Promise<ProfilerConfig> {
+// Throws error if value that must be set cannot be initialized.
+// Exported for testing purposes.
+export async function initConfig(config: Config): Promise<ProfilerConfig> {
   config = common.util.normalizeArguments(null, config);
 
-  const envConfig = {
-    logLevel: process.env.GCLOUD_PROFILER_LOGLEVEL,
+  const envConfig: any = {
     projectId: process.env.GCLOUD_PROJECT,
     serviceContext: {
       service: process.env.GAE_SERVICE,
       version: process.env.GAE_VERSION,
     }
   };
 
+  if (process.env.GCLOUD_PROFILER_LOGLEVEL !== undefined) {
+    let envLogLevel = parseInt(process.env.GCLOUD_PROFILER_LOGLEVEL || '', 10);
+    if (envLogLevel !== NaN) {
+      envConfig.logLevel = envLogLevel;
+    }
+  }
+
   let envSetConfig: Config = {};
   if (process.env.hasOwnProperty('GCLOUD_PROFILER_CONFIG')) {
     envSetConfig =
         require(path.resolve(process.env.GCLOUD_PROFILER_CONFIG)) as Config;
   }
 
-  let normalizedConfig = extend(true, {}, defaultConfig, envConfig, config);
+  let normalizedConfig = extend(
+      true, {}, defaultConfig, envSetConfig, envConfig, config, internalConfig);
 
-  if (normalizedConfig.serviceContext.service === undefined) {
-    throw new Error('service name must be specified in the configuration');
+  if (!normalizedConfig.zone || !normalizedConfig.instance) {
+    const [instance, zone] =
+        await Promise.all([getField('name'), getField('zone')])
+            .catch(
+                (err: Error) => {
+                    // ignore errors, which will occur when not on GCE.
+                }) ||
+        ['', ''];
+    if (!normalizedConfig.zone) {
+      normalizedConfig.zone = zone.substring(zone.lastIndexOf('/') + 1);
+    }
+    if (!normalizedConfig.instance) {
+      normalizedConfig.instance = instance;
+    }
   }
 
-  // TODO: fetch instance and zone from metadata. This will require function to
-  // be asynchrous.
-  if (normalizedConfig.instance === undefined) {
-    normalizedConfig.instance = '';
-  }
-  if (normalizedConfig.zone === undefined) {
-    normalizedConfig.zone = '';
+  if (normalizedConfig.serviceContext.service === undefined) {
+    throw new Error('service must be specified in the configuration');
```

Comment:
Is it possible to auto-detect the service and throw only in the case where you cannot possibly identify the serivce? Put differently, if we cannot detect the service, it is almost always our rather than the user's fault.

See the trace agent for how we deal with this.

I'd be okay if you want to do this in a follow-up, but please open an issue in that case.

Reply:
For App Engine flexible environment we should use GAE_SERVICE and GAE_VERSION for the service and service version default values respectively, similar to how the Go profiling agent does.

Reply:
We currently do use GAE_SERVICE and GAE_VERSION (in envConfig), but if GAE_SERVICE is not set, then service might not be set, and an error is thrown.
I believe this is how the go profiling agent works.